### PR TITLE
Add graph view to project overview workloads tab

### PR DESCRIPTION
**Pull request by jeff-phillips-18**
**Fixes**
https://issues.redhat.com/browse/ODC-4925

**Description**
Adds the ability to switch to the graph view in the workloads tab for the project overview page.

**Screen Shots**: 
image

image

**Browser conformance**: 
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

@bgliwa01 @bmignano @serenamarie125

### CODE REVIEW

**Comment by serenamarie125** [comment]
FYI @alimobrem @beanh66

**Comment by serenamarie125** [comment]
@jeff-phillips-18 can you verify what is happening with in-context "adds" ?

IMO, if you are in the Project->Workloads(graph) tab and do an in-context add, you would be brought to a form, and then brought back to the Project->Workloads(graph) tab to see the results.  I'm guessing right now it may navigate directly to the Topology page we use in Dev Perspective.

**Review by bgliwa01 — APPROVED** [review]
lgtm

**Review by serenamarie125 — CHANGES_REQUESTED** [review]
@jeff-phillips-18 just to be sure, I'm going to include this as a Request for Change

If you are in the Project->Workloads(graph) tab and do an in-context add, you should be brought to a form, and then brought back to the Project->Workloads(graph) tab to see the results. I'm guessing right now it may navigate directly to the Topology page we use in Dev Perspective.

**Comment by jeff-phillips-18** [comment]
@serenamarie125 currently it takes you back to the workloads tab but with the list view selected :(

**Comment by jeff-phillips-18** [comment]
@serenamarie125 fixed to redirect to the previous view for the workloads tab.

**Comment by debsmita1** [comment]
@jeff-phillips-18 In the Project workloads page, I tried adding a database using the Add to Project context menu, on creation it took me to the Template Instance Details page. I was expecting that it would bring me back to the view from where I started

File: frontend/packages/dev-console/src/components/topology/TopologyPage.tsx [hunk 1]
```diff
@@ -125,11 +113,52 @@ export const TopologyPageContext: React.FC<TopologyPageProps> = observer(({ matc
   );
 });
 
-export const TopologyPage: React.FC<TopologyPageProps> = ({ match }) => {
+type TopologyWorkloadsPageProps = {
+  match: any;
```
Comment (vikram-raj) [inline frontend/packages/dev-console/src/components/topology/TopologyPage.tsx:117]:
```suggestion
  match: RMatch<{
    name?: string;
  }>;
```

File: frontend/public/components/overview/OverviewListPage.tsx [hunk 2]
```diff
@@ -1,13 +1,10 @@
 import * as React from 'react';
-import { useParams } from 'react-router-dom';
-import DataModelProvider from '@console/dev-console/src/components/topology/data-transforms/DataModelProvider';
-import { TopologyDataRenderer } from '@console/dev-console/src/components/topology/TopologyDataRenderer';
+import TopologyPage from '@console/dev-console/src/components/topology/TopologyPage';
```
Comment (vikram-raj) [inline frontend/public/components/overview/OverviewListPage.tsx:2]: This import looks bad. We shouldn't import the dev-console package in public.

File: frontend/public/components/overview/OverviewListPage.tsx [hunk 3]
```diff
@@ -1,13 +1,10 @@
...
 
-export const OverviewListPage: React.FC = () => {
-  const namespace = useParams().name;
-  return (
-    <DataModelProvider namespace={namespace}>
-      <TopologyDataRenderer showGraphView={false} />
-    </DataModelProvider>
-  );
+type OverviewListPageProps = {
+  match: any;
```
Comment (vikram-raj) [inline frontend/public/components/overview/OverviewListPage.tsx:5]:
```suggestion
  match: RMatch<{
    ns?: string;
  }>;
```

**Comment by vikram-raj** [comment]
There is no padding around the container in the loading state. 
Screenshot 2020-10-21 at 2 56 02 PM

File: frontend/public/components/overview/OverviewListPage.tsx (hunk 2 above)
Comment (jeff-phillips-18) [reply frontend/public/components/overview/OverviewListPage.tsx:2]: This is known and already existing. I believe the plan is move Topology out of dev-console and into its own mono repo.

Comment (christianvogt) [reply frontend/public/components/overview/OverviewListPage.tsx:2]: If we are to share topology with admin console, we need to move it to a new package as a standalone tool.

**Comment by jeff-phillips-18** [comment]
@vikram-raj Updated PTAL.

File: frontend/packages/dev-console/src/components/topology/TopologyPage.tsx [hunk 4]
```diff
@@ -32,10 +38,17 @@ const getTopologyActiveView = () => {
   return localStorage.getItem(LAST_TOPOLOGY_VIEW_LOCAL_STORAGE_KEY);
 };
 
+const setTopologyWorkloadsActiveView = (id: string) => {
+  localStorage.setItem(LAST_TOPOLOGY_WORKLOADS_VIEW_LOCAL_STORAGE_KEY, id);
+};
+
+const getTopologyWorkloadsActiveView = () => {
+  return localStorage.getItem(LAST_TOPOLOGY_WORKLOADS_VIEW_LOCAL_STORAGE_KEY);
+};
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/TopologyPage.tsx:47]: Does UX want a separate setting for the topology page and for the workload page?

File: frontend/packages/dev-console/src/components/topology/TopologyPage.tsx [hunk 5]
```diff
@@ -125,11 +113,54 @@ export const TopologyPageContext: React.FC<TopologyPageProps> = observer(({ matc
...
+export const TopologyPage: React.FC<TopologyPageProps> = ({ match, workloadsPage = false }) => {
   const namespace = match.params.name;
   return (
     <DataModelProvider namespace={namespace}>
-      <TopologyPageContext match={match} />
+      {workloadsPage ? (
+        <TopologyWorkloadsPage match={match} />
+      ) : (
+        <TopologyPageContext match={match} />
+      )}
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/TopologyPage.tsx:163]: Both views are exactly the same minus the project selector.
I think they should share a lot more in common between the two pages and come up with a common strategy for the URL handling.

File: frontend/packages/operator-lifecycle-manager/src/components/operand/create-operand.tsx [hunk 6]
```diff
@@ -58,15 +59,9 @@ export const CreateOperand: React.FC<CreateOperandProps> = ({
...
   const { data: crd } = customResourceDefinition;
   const [helpText, setHelpText] = React.useState(FORM_HELP_TEXT);
-  const next =
-    activePerspective === 'dev'
-      ? '/topology'
-      : `${resourcePathFromModel(
-          ClusterServiceVersionModel,
-          match.params.appName,
-          match.params.ns,
-        )}/${match.params.plural}`;
```
Comment (christianvogt) [inline frontend/packages/operator-lifecycle-manager/src/components/operand/create-operand.tsx:68]: I don't think we can change the location we go to after creating an operand. The user who was in the `installed operators` nav item will now be redirected to the workloads tab even though an operand may not result in a workload being created.

File: frontend/packages/dev-console/src/components/topology/TopologyPage.tsx (hunk 4 above)
Comment (jeff-phillips-18) [reply frontend/packages/dev-console/src/components/topology/TopologyPage.tsx:47]: UX and PM agreed to using a single setting.

File: frontend/packages/operator-lifecycle-manager/src/components/operand/create-operand.tsx (hunk 6 above)
Comment (jeff-phillips-18) [reply frontend/packages/operator-lifecycle-manager/src/components/operand/create-operand.tsx]: Removed this change

File: frontend/packages/dev-console/src/components/topology/TopologyPage.tsx (hunk 5 above)
Comment (jeff-phillips-18) [reply frontend/packages/dev-console/src/components/topology/TopologyPage.tsx]: Updated.

File: frontend/packages/dev-console/src/components/topology/TopologyPage.tsx (hunk 4 above)
Comment (jeff-phillips-18) [reply frontend/packages/dev-console/src/components/topology/TopologyPage.tsx:46]: Sorry, missed the thought behind this. We are keeping view preference separate per sections (Workloads vs Topology). UX and PM agreed to that. The saved layout of the graph view is shared between the sections.

**Comment by serenamarie125** [comment]
@alimobrem @beanh66 FYI

**Review by jerolimov — APPROVED** [review]
I tested this a while and it works great for me.

/lgtm

File: frontend/packages/dev-console/src/components/topology/TopologyPage.tsx [hunk 7]
```diff
@@ -1,136 +1,133 @@
...
-              <CreateProjectListPage title="Topology">
-                Select a project to view the topology
-              </CreateProjectListPage>
-            )}
-          </ProjectsExistWrapper>
-        </Firehose>
-      </NamespacedPage>
-    </>
+      <div className="odc-topology-page">
+        <NamespaceBar onNamespaceChange={handleNamespaceChange} hideProjects={hideProjects}>
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/TopologyPage.tsx:96]: Why the move away from `NamespacedPage`?

Comment (jeff-phillips-18) [reply frontend/packages/dev-console/src/components/topology/TopologyPage.tsx]: Changed it back. I had tried going in another direction that is no longer relevant.

**Review by jerolimov — APPROVED** [review]
Tested it locally.
/lgtm
Happy if @christianvogt will revisit it again :)

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]
@jeff-phillips-18 Going to the workloads tab in a project without any workloads renders the Add page from developer perspective in both list and topology view. Not sure if this is the expected behavior. cc: @serenamarie125 

Peek 2020-11-04 01-15

**Comment by christianvogt** [comment]
@rohitkrai03 yes this is something we need to address with @openshift/team-devconsole-ux as the behavior exists in master already.

**Comment by jeff-phillips-18** [comment]
@jerolimov @christianvogt Updated `TopologyPage` to fix tests

**Pull request merged**